File: cpp/src/cache/cache.cpp
#include "cache.h"

namespace utilities {

template <typename KeyType, typename ValueType>
auto CacheSystem<KeyType, ValueType>::Put(const KeyType &key, const ValueType &value) -> bool {
  std::lock_guard<std::mutex> lock(cache_mutex_);

  auto it = kv_map_.find(key);
  if (it == kv_map_.end()) {
    if (kv_map_.size() >= capacity_) {
      auto victim_key = replacement_policy_->Evict();
      // evict failed
      if (!victim_key.has_value()) {
        // return false;
        throw CacheException("Evict failed");
      }
      kv_map_.erase(victim_key.value());
    }
    // add failed
    if (!replacement_policy_->Add(key)) {
      // return false;
      throw CacheException("Add key failed");
    }
    kv_map_[key] = value;
  } else {
    // update failed
    if (!replacement_policy_->Access(key)) {
      // return false;
      throw CacheException("Update key failed");
    }
    it->second = value;
  }
  return true;
}

template <typename KeyType, typename ValueType>
auto CacheSystem<KeyType, ValueType>::Get(const KeyType &key) -> std::optional<ValueType> {
  std::lock_guard<std::mutex> lock(cache_mutex_);

  auto it = kv_map_.find(key);
  // not found
  if (it == kv_map_.end()) {
    // return std::nullopt;
    throw CacheException("Key not found");
  }

  // update failed
  if (!replacement_policy_->Access(key)) {
    // return std::nullopt;
    throw CacheException("Update key failed");
  }

  return it->second;
}

template <typename KeyType, typename ValueType>
auto CacheSystem<KeyType, ValueType>::Remove(const KeyType &key) -> bool {
  std::lock_guard<std::mutex> lock(cache_mutex_);

  auto it = kv_map_.find(key);
  // not found
  if (it == kv_map_.end()) {
    // return false;
    throw CacheException("Key not found");
  }

  // remove failed
  if (!replacement_policy_->Remove(key)) {
    // return false;
    throw CacheException("Remove key failed");
  }

  kv_map_.erase(it);
  return true;
}

template class CacheSystem<int, int>;

}  // namespace utilities

File: python/cache/cache.py
# cache.py
from abc import ABC, abstractmethod
from threading import RLock
from typing import Generic, TypeVar, Optional

from cache_exception import CacheException
from replacement_policy import IReplacementPolicy
from replacement_policy_factory import ReplacementPolicyFactory, ReplacementPolicyEnum

KeyType = TypeVar("KeyType")
ValueType = TypeVar("ValueType")


class IKVCache(ABC, Generic[KeyType, ValueType]):
    """Interface for key-value cache."""

    @abstractmethod
    def put(self, key: KeyType, value: ValueType) -> bool:
        pass

    @abstractmethod
    def get(self, key: KeyType) -> Optional[ValueType]:
        pass

    @abstractmethod
    def remove(self, key: KeyType) -> bool:
        pass


class CacheSystem(IKVCache[KeyType, ValueType], Generic[KeyType, ValueType]):
    """Cache system implementing the key-value cache."""

    def __init__(
        self,
        capacity: int,
        policy_type: str,
        replacement_policy_factory: Optional[ReplacementPolicyFactory[KeyType]] = None,
    ):
        self._capacity = capacity
        self._cache = {}
        self._lock = RLock()
        factory = replacement_policy_factory or ReplacementPolicyFactory()
        self._replacement_policy = factory.create(policy_type)

    def put(self, key: KeyType, value: ValueType) -> bool:
        with self._lock:
            if key in self._cache:
                self._cache[key] = value
                if not self._replacement_policy.access(key):
                    raise CacheException("Failed to update key in replacement policy.")
            else:
                if len(self._cache) >= self._capacity:
                    victim_key = self._replacement_policy.evict()
                    if victim_key is None:
                        raise CacheException(
                            "Cache is full and no key could be evicted."
                        )
                    del self._cache[victim_key]
                if not self._replacement_policy.add(key):
                    raise CacheException("Failed to add key to replacement policy.")
                self._cache[key] = value
            return True

    def get(self, key: KeyType) -> Optional[ValueType]:
        with self._lock:
            if key not in self._cache:
                return None
            if not self._replacement_policy.access(key):
                raise CacheException("Failed to access key in replacement policy.")
            return self._cache[key]

    def remove(self, key: KeyType) -> bool:
        with self._lock:
            if key not in self._cache:
                return False
            if not self._replacement_policy.remove(key):
                raise CacheException("Failed to remove key from replacement policy.")
            del self._cache[key]
            return True
